test/c/test_sparse_observable.c

```diff
@@ -0,0 +1,650 @@
+
+    qk_obs_free(obs);
+    qk_obs_free(simplified);
+
+    if (num_terms != 2 || num_terms_simplified != 1) {
+        return EqualityError;
+    }
+    return Ok;
+}
+
+/**
+ * Test getting the terms in an observable.
+ */
+int test_term() {
+    uint32_t num_qubits = 100;
+    QkSparseObservable *obs = qk_obs_identity(num_qubits);
+
+    QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    uint32_t qubits[3] = {0, 1, 2};
+    complex double coeff = 1 + I;
+
+    QkSparseTerm term = {coeff, 3, bit_terms, qubits, num_qubits};
+    int err = qk_obs_add_term(obs, &term);
+
+    if (err != 0) {
+        return RuntimeError;
+    }
+
+    // some placeholders to store the results
+    int nnis[2] = {-1, -1};
+    int bits[3] = {-1, -1, -1};
+    int indices[3] = {-1, -1, -1};
+
+    uint64_t num_terms = qk_obs_num_terms(obs);
+    for (uint64_t i = 0; i < num_terms; i++) {
+        QkSparseTerm view;
+        qk_obs_term(obs, i, &view);
+        size_t nni = view.len;
+        nnis[i] = nni; // store to compare later
+
+        for (uint32_t n = 0; n < nni; n++) {
+            // this loop is only called once, so we can use ``n`` to index here
+            bits[n] = view.bit_terms[n];
+            indices[n] = view.indices[n];
+        }
+    }
+
+    qk_obs_free(obs);
+
+    int result = Ok;
+    int expected_nnis[2] = {0, 3};
+    int expected_bits[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    int expected_indices[3] = {0, 1, 2};
+
+    // check number of terms
+    if (num_terms != 2) {
+        result = EqualityError;
+    }
+
+    // check NNIs
+    for (int i = 0; i < 2; i++) {
+        if (nnis[i] != expected_nnis[i]) {
+            result = EqualityError;
+        }
+    }
+
+    // check bit terms and indices
+    for (int n = 0; n < 3; n++) {
+        if (indices[n] != expected_indices[n] || bits[n] != expected_bits[n]) {
+            result = EqualityError;
+        }
+    }
+
+    return result;
+}
+
+/**
+ * Test copying and modifying a term.
+ */
+int test_copy_term() {
+    // create an observable with the term X0 Y1 Z2
+    u_int32_t num_qubits = 100;
+    QkSparseObservable *obs = qk_obs_zero(num_qubits);
+
+    complex double coeff = 1;
+    QkBitTerm bits[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    uint32_t indices[3] = {0, 1, 2};
+
+    QkSparseTerm term = {coeff, 3, bits, indices, num_qubits};
+    qk_obs_add_term(obs, &term);
+
+    // now we add a modified copy of the first term -- we use use qk_obs_term(..., &borrowed) on
+    // purpose here
+    QkSparseTerm borrowed;
+    int error = qk_obs_term(obs, 0, &borrowed); // get view on 0th term
+    if (error > 0) {
+        return RuntimeError;
+    }
+
+    // copy the term so we can safely modify it and add it onto the observable
+    size_t len = borrowed.len;
+    QkBitTerm *copied_bits = (QkBitTerm *)malloc(len * sizeof(QkBitTerm));
+    uint32_t *copied_indices = (uint32_t *)malloc(len * sizeof(uint32_t));
```

Comment:
Given we know the expected sizes here, if the dynamic memory allocation gets too fiddly with cleanups, etc, we could move towards patterns where we declare `QkBitTerm copied_bits[4]` or so on the stack, then assert that `borrowed.len` is what we expect, and then we don't need to bother with `free` because it's stack-allocated.

Just commenting for consideration in future testing patterns.